bluetooth/src/commonMain/kotlin/device/Device.kt

```diff
@@ -0,0 +1,323 @@
+                    is BaseDeviceConnectionManager.Event.Disconnected,
+                    is BaseDeviceConnectionManager.Event.MtuUpdated -> deviceStateRepo.value
+                }
+                repo?.takeAndChangeState { state ->
+                    event.stateTransition(state)
+                }
+            }
+        }
+    }
+
+    override suspend fun connect(): Boolean = createDeviceStateRepoIfNotCreated()?.let { repo ->
+        repo.transformLatest { deviceState ->
+            when (deviceState) {
+                is ConnectibleDeviceState.Disconnected -> deviceState.startConnecting()
+                is ConnectibleDeviceState.Connected -> emit(true)
+                is ConnectibleDeviceState.Connecting, is ConnectibleDeviceState.Reconnecting, is ConnectibleDeviceState.Disconnecting -> {}
+            }
+        }.first()
+    } ?: false
+
+    override fun handleConnected() = createConnectionManagerIfNotCreated().handleConnect()
+
+    override suspend fun disconnect() {
+        deviceStateRepo.value?.transformLatest { deviceState ->
+            when (deviceState) {
+                is ConnectibleDeviceState.Connected -> deviceState.startDisconnected()
+                is ConnectibleDeviceState.Connecting -> deviceState.handleCancel()
+                is ConnectibleDeviceState.Reconnecting -> deviceState.handleCancel()
+                is ConnectibleDeviceState.Disconnected -> emit(Unit)
+                is ConnectibleDeviceState.Disconnecting -> {} // just wait
+            }
+        }?.first()
+    }
+
+    override suspend fun handleDisconnected() = createConnectionManagerIfNotCreated().handleDisconnect()
+
+    override fun rssiDidUpdate(rssi: Int) {
+        sharedInfo.value = sharedInfo.value.copy(rssi = rssi)
+    }
+
+    override fun advertisementDataAndRssiDidUpdate(
+        advertisementData: BaseAdvertisementData,
+        rssi: Int
+    ) {
+        sharedInfo.value = sharedInfo.value.copy(rssi = rssi, advertisementData = advertisementData)
+    }
+
+    private fun createConnectionManagerIfNotCreated(): BaseDeviceConnectionManager = if (connectionManager.isCompleted) {
+        connectionManager.getCompleted()
+    } else {
+        connectionManagerBuilder().also {
+            connectionManager.complete(it)
+        }
+    }
+
+    private fun createDeviceStateRepoIfNotCreated(): ConnectibleDeviceStateFlowRepo? =
+        deviceStateRepo.updateAndGet { repo ->
+            repo ?: if (sharedInfo.value.advertisementData.isConnectable) createDeviceStateFlow(
+                createConnectionManagerIfNotCreated(),
+                coroutineScope.coroutineContext
+            ) else null
+        }
+
+    private suspend fun BaseDeviceConnectionManager.Event.stateTransition(state: ConnectibleDeviceState): suspend () -> ConnectibleDeviceState =
+        when (this) {
+            is BaseDeviceConnectionManager.Event.Connecting -> stateTransition(state)
```

Comment:
Each case does the same, isn't it?

Reply:
No, its not a cyclical loop. The individual events have their own stateTransition implementation to handle their internal state transition behaviour.

Reply:
Got it right after comment by scrolling down